### ISSUE TITLE
Fails to load plugin_python.dll when launched from desktop on linux

### ISSUE THREAD

**Issue by foresto** [ModOrganizer2/modorganizer#1664]
## The problem:

When running on linux (Wine) from a desktop menu instead of a terminal window, the Python plugin fails to initialize. Relevant error message:

    Fatal Python error: init_sys_streams: can't initialize sys standard streams
    OSError: [WinError 6] Invalid handle

Mod Organizer then exits without ever showing the GUI. Trying to launch it again brings up a "failed to load the plugin" dialog box, asking whether to skip or blacklist the Python plugin.

I have only seen this happen when launching ModOrganizer directly from a .desktop launcher or desktop environment menu, where the process is launched without an attached terminal (or perhaps without a TERM environment variable). My guess is that the Windows Python plugin has trouble figuring out the character encodings for stdin/stdout/stderr streams when run in this environment, and just gives up.

## Workarounds:

Conveniently, most linux systems have python natively installed, and python's standard library can run another process with a pseudo-terminal attached. We can use this to work around the problem.

Instead of running a command line like this:

    wine ModOrganizer.exe

Use a command line like this:

    python3 -c "import pty; pty.spawn( ['wine', 'ModOrganizer.exe'] )"

## Notes:

I do not expect support on this issue. I'm filing it for the sake of documentation, in case another linux user encounters it, since the cause and workaround are fairly arcane.

I imagine it might be possible for ModOrganizer's Python plugin to solve the problem by doing a little cross-platform work in its setup phase. I don't have time to investigate this myself, though.

I'm using OS containers, so it's possible that this failure only occurs in container-like restricted environments. I haven't tested it outside of my particular container environment.

I'll reference #372 here, even though this is not USVFS related, for visibility to linux users.

**Comment by Nuanulla** [comment]
Can this Python line be included in Steam’s launch options? I have issue with the Python plugin on Steam Deck, but only in Game Mode. I’ve blacklisted the plug-in for now. Running MO 2.5.0 directly through Steam.

**Comment by Nuanulla** [comment]
Unfortunately, `</dev/null` didn't work for me. What I instead did was adapt your python line for Steam Proton as shown below:
`STEAM_COMPAT_DATA_PATH="/home/deck/.local/share/Steam/steamapps/compatdata/489830" %command% python3 -c "import pty; pty.spawn( ['$HOME/.local/share/Steam/steamapps/common/Proton 7.0/proton', '/home/deck/.local/share/Steam/steamapps/compatdata/489830/pfx/drive_c/Modding/MO2/ModOrganizer.exe'] )"`

This does start MO2, but it now states "Cannot start Python 3", followed by "The file 'C:\Modding\MO2\python3' does not exist." Any ideas where to go next?

**Comment by Nuanulla** [comment]
Nvm. I had a couple fundamental misunderstandings, so I only recently realized the following:
- lines after %command% are passed to the program as arguments
- Steam Deck cannot access certain directories and commands when in Game Mode
- MO2 is bundled with Python 3.10 in MO 2.5.0 alpha 6. 

The above explain why I was going in the wrong direction. I was able to follow your suggestion and use the script command in Steam, through a .desktop file, and in terminal by defining env and and compatdata folder. However, as mentioned, Deck unfortunately cannot use it outside of Desktop. I'm just avoiding STL and Linux MO2 Installer out of both interest in setting MO2 up myself and distaste for their respective implementations. If plugin_python is disabled on my current setup, I actually have no issues running MO2 as it is but without plugin error checking.

Given MO2 works perfectly find in Desktop Mode on Deck, this means that something is prohibiting MO2 from accessing its own python310.dll when Deck is in Game Mode.

**Comment by Nuanulla** [comment]
> It appears there's a fundamental problem with the way ModOrganizer loads its python interpreter. It shouldn't fail just because the process has no standard input/output streams, but it does. Given how widely used python is in many different environments, I suspect there's a way to fix whatever ModOrganizer is doing, but I haven't looked at the code that does it.

At the same time, I've successfully used MO2 2.4.4 through STL and ModOrganizer2 Linux Installer, and both are able to run MO2's python interpreter just fine. I feel like I must be missing something obvious - either an easy fix, or an obvious reason why my current MO2 setup won't ever launch the python interpreter correctly with the way MO2 itself currently is. I've already tried MO2 2.4.4  which, when used outside of STL and the linux installer, is unable to run the python interpreter successfully much like 2.5.0 alpha 6.

**Comment by Holt59** [comment]
You have to understand that your external Python installation (on Steam deck or whatever) as nothing to do with the MO2 python plugin. MO2 ships its "own" Python interpreter through the Python DLL, however there is a major difference between 2.4 and 2.5:

- MO2 2.4 uses Boost.Python and targets Python 3.8,
- MO2 2.5 uses pybind11 and targets Python 3.10.

Also, MO2 2.5 has a completely different plugin structure for the Python plugins, which may create different behaviors on your side.

**Comment by Nuanulla** [comment]
> You have to understand that your external Python installation (on Steam deck or whatever) as nothing to do with the MO2 python plugin.

I had established above that I've now realized that fact.

> however there is a major difference between 2.4 and 2.5

Unfortunately, and also as we've already mentioned in the exchange above, the MO2 version (whether 2.4 or 2.5 alpha) doesn't make any difference. It appears to have nothing to do with whatever python version. Rather, for my case at least, it's to do with the way MO2 initializes Python which is interfered with by Steam Deck's Game Mode. Yet, STL and the Linux Installer show no issues in this aspect.

**Comment by Holt59** [comment]
> 
> > however there is a major difference between 2.4 and 2.5
> 
> Unfortunately, and also as we've already mentioned in the exchange above, the MO2 version (whether 2.4 or 2.5 alpha) doesn't make any difference. It appears to have nothing to do with whatever python version. Rather, for my case at least, it's to do with the way MO2 initializes Python which is interfered with by Steam Deck's Game Mode. Yet, STL and the Linux Installer show no issues in this aspect.

Unfortunately, that's an issue with the Python interpreter itself, not with MO2. MO2 does not initialize the interpreter in any special way (at least up to the point where your error occurs), you can check https://github.com/ModOrganizer2/modorganizer-plugin_python/blob/26d7de23f34c9cf1b59500453ae40522199715fc/src/runner/pythonrunner.cpp#L96-L110

You can find instances of this issue for other programs with kind of similar setups, e.g., docker, etc., but that's most likely to be fixed by Python itself.

**Comment by Nuanulla** [comment]
> However, I ran into a similar problem with another tool, where adding `</dev/null` to the end of the Steam launch options worked. I don't know if it would help in this case.

I tried to debug in Game Mode this time by writing out to an output.log with `&> $HOME/debug.log`, but this appeared to make the plugin_python.dll load properly in Game Mode. So, I took a second harder look at your prior suggestion and came up with the following launch argument:
`STEAM_COMPAT_DATA_PATH="/home/deck/.local/share/Steam/steamapps/compatdata/489830" %command% &> /dev/null`

It's like I said: I was missing something obvious, out of my inexperience. This would explain why rockerbacon's MO2 Linux Installer works: because, in Game Mode, it continues outputting to a log. Here, as I understand, I'm outputting to a void with `/dev/null`. Could you explain to me why you typed `</dev/null` instead of `> /dev/null`, and why this method appears to help MO2 find its Python .dll?

**Comment by Holt59** [comment]
> BTW, I think this would be relatively easy to fix for someone who has a MO build environment and understanding of file stream handling: Check file descriptors 0, 1, and 2, and if they aren't open already, open them (connected to the logger or the controlling terminal or /dev/null as appropriate) before starting the Python plugin.

As already said, this not a MO2 fix, this is a Python fix. On the MO2 side, we have no idea what Python is trying to open or check in term of handle, we simply call the initialization function from the interpreter, so there is no "check file descriptors". Digging into this means digging into the CPython implementation itself, trying to fix internal stuff, which is a very bad idea as it could completely breaks for future release of Python.

**Comment by Holt59** [comment]
> @Holt59, you're misunderstanding a lot here, both technically and socially, but I'm not interested in arguing with you about it.

I don’t think so. Technically I basically wrote the Python plugin twice and I’m working on a daily basis with Python/C++ interface. Socially, well you just proved my point.

> In any case, I don't expect or need a fix for myself. I've already done the hard work on this issue, by discovering, analyzing, and documenting the most likely cause, as well as finding multiple workarounds. I hope it's helpful to the next person to be bitten by the bug.

**Comment by geojoek** [comment]
Any insight on why it might be doing similar behavior with exiting before GUI, but staying on as a background process, but on Windows 10 ([Issue #1672](https://github.com/ModOrganizer2/modorganizer/issues/1672))?  If I kill the process, the next time I load MO2 it gives me the plugin_python.dll failed-to-access error.  I also have it loaded on an M.2 nvme as part of the whole Nolvus Skyrim expansion pack.
